nuget/rules.bzl

```diff
@@ -0,0 +1,332 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# This file is based on the original implementation of https://github.com/SeleniumHQ/selenium/.
+
+load("@bazel_skylib//rules:common_settings.bzl", "BuildSettingInfo")
+load("@rules_dotnet//dotnet/private:common.bzl", "is_debug")
+load("@rules_dotnet//dotnet/private:providers.bzl", "DotnetAssemblyRuntimeInfo")
+
+# The change to the PATH is recommended here:
+# https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual?source=recommendations#set-environment-variables-system-wide
+# We list our .Net installation first because we
+# want it to be picked up first
+
+# The `MSBuildEnableWorkloadResolver` is disabled to prevent warnings
+# about a missing Microsoft.NET.SDK.WorkloadAutoImportPropsLocator
+
+def dotnet_preamble(toolchain):
+    return """
+export DOTNET="$(pwd)/{dotnet}"
+export DOTNET_CLI_HOME="$(dirname $DOTNET)"
+export DOTNET_CLI_TELEMETRY_OPTOUT=1
+export DOTNET_NOLOGO=1
+export DOTNET_ROOT="$(dirname $DOTNET)"
+export PATH=$DOTNET_ROOT:$DOTNET_ROOT/tools:$PATH
+export MSBuildEnableWorkloadResolver=false
+export CWD=$(pwd)
+
+# Required to make packing work on Windows
+export APPDATA="$(pwd)"
+export PROGRAMFILES="$(pwd)"
+
+# Required to make NuGet tool work on non-writable home path like GitHub actions
+export XDG_DATA_HOME=$(mktemp -d)
+
+# Create `global.json` to trick .Net into using the hermetic toolchain
+# https://learn.microsoft.com/en-us/dotnet/core/tools/global-json
+echo '{{"sdk": {{"version": "{version}"}} }}' >$(pwd)/global.json
+
+""".format(
+        dotnet = toolchain.runtime.files_to_run.executable.path,
+        version = toolchain.dotnetinfo.sdk_version,
+    )
+
+
+def _check_platform(platform):
+    allowed_values = ("osx-arm64", "osx-x64", "linux-arm64", "linux-x64", "win-arm64", "win-x64")
+    if platform not in allowed_values:
+        fail("Platform must be set to any of {}. Got {} instead!".format(allowed_values, platform))
+
+
+def _parse_version(ctx):
+    version = ctx.attr.version
+    if not version:
+        version = ctx.var.get("version", "0.0.0")
+
+    return version
+
+
+def _nuget_pack_impl(ctx):
```

Comment:
Initially taken from [here](https://github.com/SeleniumHQ/selenium/blob/cca70e3eb49c2230739d10f828f99252b913fda8/dotnet/private/nuget_pack.bzl), but this version has many small modifications.